akka-sample-persistence-dc-scala/src/main/scala/sample/persistence/multidc/ThumbsUpHttp.scala

```diff
@@ -1,52 +1,46 @@
+import akka.actor.typed.ActorSystem
+import akka.cluster.sharding.typed.{ReplicatedSharding, ShardingEnvelope}
 import akka.http.scaladsl.Http
-import akka.http.scaladsl.model.ContentTypes
-import akka.http.scaladsl.model.HttpEntity
-import akka.http.scaladsl.model.HttpResponse
-import akka.http.scaladsl.model.StatusCodes
-import akka.pattern.ask
-import akka.stream.ActorMaterializer
+import akka.http.scaladsl.model.{ContentTypes, HttpEntity, HttpResponse, StatusCodes}
+import akka.persistence.typed.ReplicaId
 import akka.stream.scaladsl.Source
-import akka.util.ByteString
-import akka.util.Timeout
+import akka.util.{ByteString, Timeout}
+import sample.persistence.multidc.ThumbsUpCounter.State
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
 
 object ThumbsUpHttp {
 
-  def startServer(httpHost: String, httpPort: Int, counterRegion: ActorRef)(implicit system: ActorSystem) = {
+  def startServer(httpHost: String, httpPort: Int, selfReplica: ReplicaId, res: ReplicatedSharding[ThumbsUpCounter.Command])(implicit system: ActorSystem[_]): Unit = {
 
     import akka.http.scaladsl.server.Directives._
 
-    implicit val mat = ActorMaterializer()
-    import system.dispatcher
     implicit val timeout: Timeout = Timeout(10.seconds)
+    implicit val ec: ExecutionContext = system.executionContext
 
     val api = pathPrefix("thumbs-up") {
       concat(
         // example: curl http://127.0.0.1:22551/thumbs-up/a
         get {
-          path(Segment) { resourceId ⇒
-            onComplete((counterRegion ? ThumbsUpCounter.GetUsers(resourceId))
-              .mapTo[ThumbsUpCounter.State]) { state =>
-              val s = Source.fromIterator(() ⇒ state.get.users.iterator)
-                .intersperse("\n")
-                .map(ByteString(_))
-
-              complete(HttpResponse(entity = HttpEntity(ContentTypes.`text/plain(UTF-8)`, s)))
+          path(Segment) { resourceId =>
+            onComplete(res.entityRefsFor(resourceId)(selfReplica).ask[State](replyTo => ThumbsUpCounter.GetUsers(resourceId, replyTo))) {
+              case Success(state) =>
+                val s = Source.fromIterator(() => state.users.iterator)
+                  .intersperse("\n")
+                  .map(ByteString(_))
+                complete(HttpResponse(entity = HttpEntity(ContentTypes.`text/plain(UTF-8)`, s)))
+              case Failure(ex) => complete(StatusCodes.InternalServerError, ex.toString)
             }
           }
         },
         // example: curl -X POST http://127.0.0.1:22551/thumbs-up/a/u1
         post {
-          path(Segment / Segment) { (resourceId, userId) ⇒
-            onComplete((counterRegion ? ThumbsUpCounter.GiveThumbsUp(resourceId, userId)).mapTo[Int]) {
+          path(Segment / Segment) { (resourceId, userId) =>
+            onComplete(res.entityRefsFor(resourceId)(selfReplica).ask[Long](replyTo => ThumbsUpCounter.GiveThumbsUp(resourceId, userId, replyTo))) {
```

Comment:
I could see that it would be nice with a convenience for getting the selfReplica ref, but that can be added later.